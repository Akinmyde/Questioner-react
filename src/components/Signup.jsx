import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { toast } from 'react-toastify';
import Loader from './common/Loader';
import Input from './common/Input';
import Button from './common/Button';
import validate from '../helpers/validator';
import signupSchema from '../schema/signupSchema';
import { register } from './services/authService';
import exceptionHandler from '../helpers/exceptionHandler';

class Signup extends Component {
  constructor(props) {
    super(props);
    this.state = {
      account: {
        username: '', email: '', password: '', confirmPassword: '',
      },
      loading: false,
    };
  }

  render() {
    const { account: userAcount, loading } = this.state;
    const {
      username, email, password, confirmPassword,
    } = userAcount;

    const doSumbit = async () => {
      // call the server
      this.setState({ loading: true });
      try {
        await register(userAcount);
      } catch (ex) {
        exceptionHandler(ex);
      } finally {
        this.setState({ loading: false });
      }
    };

    const handelClick = (e) => {
      e.preventDefault();
      const error = validate(userAcount, signupSchema);
      if (error) return toast.error(error);
      return doSumbit();
    };

    const hanleChange = ({ currentTarget: input }) => {
      const account = { ...userAcount };
      account[input.id] = input.value;
      this.setState({ account });
    };

    return (
      <React.Fragment>
        {loading && <Loader />}
        <div className="container">
          <div className="imgcontainer">
            <i className="fas fa-user-circle fa-5x" />
          </div>
          <Input
            onChange={hanleChange}
            autoFocus="autoFocus"
            value={email}
            type="email"
            id="email"
            label="Email"
            placeholder="johndeo@gmail.com"
            required="required"
          />
          <Input
            onChange={hanleChange}
            value={username}
            type="text"
            id="username"
            label="Username"
            placeholder="John Deo"
            required="required"
          />
          <Input
            onChange={hanleChange}
            value={password}
            type="password"
            id="password"
            label="Password"
            placeholder="password"
            required="required"
          />
          <Input
            onChange={hanleChange}
            value={confirmPassword}
            type="password"
            id="confirmPassword"
            label="Retype Password"
            placeholder="password"
            required="required"
          />
          <div className="right font12"><Link to="/forget">Forget Password</Link></div>
          <Button onClick={handelClick} id="Signup" value="Signup" />
          <div className="center font12">
            {'Already have an account?'}
            {' '}
            <Link to="/login">Login</Link>
          </div>
          <br />
          <div className="error font12" />
        </div>
      </React.Fragment>
    );
  }
}

export default Signup;
