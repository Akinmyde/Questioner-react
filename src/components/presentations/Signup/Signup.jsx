import React from 'react';
import { Link, Redirect } from 'react-router-dom';
import Form from '../../common/Form';
import Loader from '../../common/Loader';
import signupSchema from '../../../schema/signupSchema';

class Signup extends Form {
  constructor(props) {
    super(props);
    this.state = {
      data: {
        username: '', email: '', password: '', confirmPassword: '',
      },
      loading: false,
    };
  }

  schema = { ...signupSchema };

  doSubmit = async () => {
    const { data } = this.state;
    const { signup } = this.props;
    await signup(data, this.props);
  };

  render() {
    const { auth, LoadingReducer } = this.props;
    const { userId } = auth;
    const { loader } = LoadingReducer;

    if (userId) return <Redirect to="/" />;
    return (
      <React.Fragment>
        {loader && <Loader />}
        <div className="container">
          <div className="imgcontainer">
            <i className="fas fa-user-circle fa-5x" />
          </div>
          {this.renderInput('email', 'Email', 'JohnDoe@gmail.com', 'email')}
          {this.renderInput('username', 'Username', 'John Doe')}
          {this.renderInput('password', 'Password', 'password', 'password')}
          {this.renderInput('confirmPassword', 'Retype Password', 'Confirm Password', 'password')}
          <div className="right font12"><Link to="/forget">Forget Password</Link></div>
          {this.renderButton('Signup')}
          <div className="center font12">
            {'Already have an account?'}
            {' '}
            <Link to="/login">Login</Link>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export default Signup;
