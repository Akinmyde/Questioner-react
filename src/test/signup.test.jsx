import 'babel-polyfill';
import React from 'react';
import { shallow } from 'enzyme';
import Signup from '../components/Signup';

describe('Signup Page', () => {
  it('should render signup correctly', () => {
    const signup = shallow(<Signup />);
    expect(signup).toMatchSnapshot();
  });

  it('should test functions', () => {
    const wrapper = shallow(<Signup />);
    expect(wrapper.instance().doSubmit()).toMatchSnapshot();
  });
});
