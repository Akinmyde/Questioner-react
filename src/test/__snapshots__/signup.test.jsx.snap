// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Page should render signup correctly 1`] = `
<Fragment>
  <div
    className="container"
  >
    <div
      className="imgcontainer"
    >
      <i
        className="fas fa-user-circle fa-5x"
      />
    </div>
    <Input
      autoFocus=""
      id="email"
      label="Email"
      onChange={[Function]}
      placeholder="JohnDoe@gmail.com"
      type="email"
      value=""
    />
    <Input
      autoFocus=""
      id="username"
      label="Username"
      onChange={[Function]}
      placeholder="John Doe"
      type="text"
      value=""
    />
    <Input
      autoFocus=""
      id="password"
      label="Password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <Input
      autoFocus=""
      id="confirmPassword"
      label="Retype Password"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value=""
    />
    <div
      className="right font12"
    >
      <Link
        to="/forget"
      >
        Forget Password
      </Link>
    </div>
    <Button
      className="btn"
      onClick={[Function]}
      value="Signup"
    />
    <div
      className="center font12"
    >
      Already have an account?
       
      <Link
        to="/login"
      >
        Login
      </Link>
    </div>
  </div>
</Fragment>
`;
